src/loaders/FileLoader.js

```diff
@@ -55,225 +53,178 @@ class FileLoader extends Loader {
+				const response = new Response( stream );
 
-					}
+				switch ( this.responseType ) {
 
-					scope.manager.itemEnd( url );
+					case 'arraybuffer':
 
-				} else {
+						return response.arrayBuffer();
 
-					for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
+					case 'blob':
 
-						const callback = callbacks[ i ];
-						if ( callback.onError ) callback.onError( event );
+						return response.blob();
 
-					}
+					case 'document':
 
-					scope.manager.itemError( url );
-					scope.manager.itemEnd( url );
+						return response.text()
+							.then( text => {
 
-				}
+								const parser = new DOMParser();
+								return parser.parseFromString( text, this.mimeType );
 
-			}, false );
+							} );
 
-			request.addEventListener( 'progress', function ( event ) {
+					case 'json':
 
-				const callbacks = loading[ url ];
+						return response.json();
 
-				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
+					default:
 
-					const callback = callbacks[ i ];
-					if ( callback.onProgress ) callback.onProgress( event );
+						return response.text();
 
 				}
 
-			}, false );
+			} )
+			.then( data => {
 
-			request.addEventListener( 'error', function ( event ) {
+				// Add to cache only on HTTP success, so that we do not cache
+				// error response bodies as proper responses to requests.
+				Cache.add( url, data );
 
 				const callbacks = loading[ url ];
-
 				delete loading[ url ];
 
 				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
 					const callback = callbacks[ i ];
-					if ( callback.onError ) callback.onError( event );
+					if ( callback.onLoad ) callback.onLoad( data );
 
 				}
 
-				scope.manager.itemError( url );
-				scope.manager.itemEnd( url );
+				this.manager.itemEnd( url );
 
-			}, false );
+			} )
+			.catch( err => {
 
-			request.addEventListener( 'abort', function ( event ) {
+				// Abort errors and other errors are handled the same
 
 				const callbacks = loading[ url ];
-
 				delete loading[ url ];
 
 				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
 					const callback = callbacks[ i ];
-					if ( callback.onError ) callback.onError( event );
+					if ( callback.onError ) callback.onError( err );
 
 				}
 
-				scope.manager.itemError( url );
-				scope.manager.itemEnd( url );
+				this.manager.itemError( url );
+				this.manager.itemEnd( url );
 
-			}, false );
-
-			if ( this.responseType !== undefined ) request.responseType = this.responseType;
-			if ( this.withCredentials !== undefined ) request.withCredentials = this.withCredentials;
-
-			if ( request.overrideMimeType ) request.overrideMimeType( this.mimeType !== undefined ? this.mimeType : 'text/plain' );
-
-			for ( const header in this.requestHeader ) {
-
-				request.setRequestHeader( header, this.requestHeader[ header ] );
-
-			}
-
-			request.send( null );
-
-		}
+			} );
 
-		scope.manager.itemStart( url );
+		this.manager.itemStart( url );
 
-		return request;
+		return;
```

Comment:
nit: there's no longer a need for this `return` statement

Reply:
sure, can do